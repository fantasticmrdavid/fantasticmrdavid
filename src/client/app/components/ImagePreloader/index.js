import React, { Component } from 'react';

export default class ImagePreloader extends Component {
  constructor (props) {
    super(props);

    const { images } = props;

    this.state = {
      count: images.length,
      completed: 0,
    };

    this.loaders = [];
  }

  componentWillMount () {
    const { images, completedAction } = this.props;

    images.map((i) => {
      let image = new Image;

      image.onload = () => {
        let { completed } = this.state;
        this.setState({
          completed: parseInt(completed) + 1,
        });
        this.checkCompleted() && completedAction();
      };
      image.src = i;
      this.loaders.push(image);
    });
  }

  checkCompleted () {
    const { count, completed } = this.state;
    return count === completed;
  }

  render () {
    return null;
  }
}
