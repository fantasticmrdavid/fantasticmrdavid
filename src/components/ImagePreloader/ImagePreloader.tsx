import React, {
  memo,
  useCallback,
  useEffect,
  useReducer
} from "react";
import { Container } from './imagePreloader.styles';

type State = {
  completed: boolean,
  completedCount: number,
};

type Action = {
  type: string,
};

const reducer = (state: State, action: Action) => {
  switch (action.type) {
    case 'incrementCompleted':
      return {
        ...state,
        completedCount: state.completedCount + 1,
      };
    case 'completeLoading':
      return {
        ...state,
        completed: true,
      };
    default: throw new Error();
  }
};

interface ImagePreloaderProps {
  images: string[],
  hideCompletedCount?: boolean,
  completedAction?: (...args: any[]) => any,
}

export const ImagePreloader = memo(
  ({ images, completedAction, hideCompletedCount }: ImagePreloaderProps) => {
    const [state, dispatch] = useReducer(reducer, {
      completedCount: 0,
      completed: false,
    });
    const initImage = useCallback((src: string) => {
      const image = new Image();

      image.onload = () => {
        if (!state.completed) dispatch({ type: 'incrementCompleted' });
      };
      image.src = src;
    }, [state.completedCount]);

    images.map((i) => useEffect(() => initImage(i), [i]));

    useEffect(() => {
      if (!state.completed && state.completedCount === images.length) {
        if (completedAction) completedAction();
        dispatch({ type: 'completeLoading' });
      }
    });

    return (
      <Container completed={state.completed}>
        { !hideCompletedCount && <>{(state.completedCount / images.length) * 100}%</> }
      </Container>
    );
  },
);
