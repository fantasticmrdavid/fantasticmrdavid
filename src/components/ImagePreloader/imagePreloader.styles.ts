import styled, { css } from 'styled-components';
import theme from 'styles/theme';
import { confirmReset } from 'styles/animations';

interface ContainerProps {
  $isCompleted: boolean,
}

export const Container = styled.div<ContainerProps>`
  position: relative;
  font-size: 1.25rem;
  font-weight: 700;
  color: ${theme.colorWhite};
  text-align: center;
  animation: ${(props) => (props.$isCompleted ? css`${confirmReset} 1s normal linear` : undefined)};
  animation-fill-mode: forwards;
`;
