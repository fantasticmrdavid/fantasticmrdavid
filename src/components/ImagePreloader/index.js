import { PureComponent } from 'react';
import PropTypes from 'prop-types';

export default class ImagePreloader extends PureComponent {
  constructor(props) {
    super(props);

    const { images } = props;

    this.state = {
      count: images.length,
      completed: 0,
    };

    this.loaders = [];
  }

  componentWillMount() {
    const { images, completedAction } = this.props;

    images.map((i) => {
      const image = new Image();

      image.onload = () => {
        const { completed } = this.state;
        this.setState({
          completed: parseInt(completed, 10) + 1,
        });
        if (this.checkCompleted()) completedAction();
      };
      image.src = i;
      this.loaders.push(image);

      return i;
    });
  }

  checkCompleted() {
    const { count, completed } = this.state;
    return count === completed;
  }

  render() {
    return null;
  }
}

ImagePreloader.propTypes = {
  images: PropTypes.array.isRequired,
  completedAction: PropTypes.func,
};
