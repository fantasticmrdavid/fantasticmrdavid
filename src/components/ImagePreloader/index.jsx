import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import * as styles from './styles';

export default class ImagePreloader extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      completed: 0,
    };

    this.loaders = [];
  }

  componentWillMount() {
    const { images, completedAction } = this.props;

    images.map((i) => {
      const image = new Image();

      image.onload = () => {
        const { completed } = this.state;
        this.setState({
          completed: parseInt(completed, 10) + 1,
        });
        if (this.checkCompleted()) completedAction();
      };
      image.src = i;
      this.loaders.push(image);

      return i;
    });
  }

  checkCompleted() {
    const { images } = this.props;
    const { completed } = this.state;
    return images.length === completed;
  }

  render() {
    const { images } = this.props;
    const { completed } = this.state;
    const { Container } = styles;
    return (
      <Container completed={completed === images.length}>
        {completed / images.length * 100}%
      </Container>
    );
  }
}

ImagePreloader.propTypes = {
  images: PropTypes.array.isRequired,
  completedAction: PropTypes.func,
};
