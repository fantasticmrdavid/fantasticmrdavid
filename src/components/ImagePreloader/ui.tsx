import React, {
  memo,
  useCallback,
  useEffect,
  useReducer,
} from 'react';
import * as styles from './styles';

type State = {
  completed: boolean,
  completedCount: number,
};

type Action = {
  type: string,
};

const reducer = (state: State, action: Action) => {
  switch (action.type) {
    case 'incrementCompleted':
      return {
        ...state,
        completedCount: state.completedCount + 1,
      };
    case 'completeLoading':
      return {
        ...state,
        completed: true,
      };
    default: throw new Error();
  }
};

interface Props {
  images: string[],
  completedAction?: (...args: any[]) => any,
}

const { Container } = styles;

export default memo(
  ({ images, completedAction }: Props) => {
    const [state, dispatch] = useReducer(reducer, {
      completedCount: 0,
      completed: false,
    });
    const initImage = useCallback((src: string) => {
      const image = new Image();

      image.onload = () => {
        if (!state.completed) dispatch({ type: 'incrementCompleted' });
      };
      image.src = src;
    }, [state.completedCount]);

    images.map((i) => useEffect(() => initImage(i), [i]));

    useEffect(() => {
      if (!state.completed && state.completedCount === images.length) {
        if (completedAction) completedAction();
        dispatch({ type: 'completeLoading' });
      }
    });

    return (
      <Container completed={state.completed}>
        {(state.completedCount / images.length) * 100}%
      </Container>
    );
  },
);
