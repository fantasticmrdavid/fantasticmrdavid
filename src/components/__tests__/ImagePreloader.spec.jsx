import React from 'react';
import renderer from 'react-test-renderer';
import ImagePreloader from 'components/ImagePreloader';

const images = [
  '/assets/images/legoFlinders.jpg',
];

const completedAction = jest.fn();

describe('Component -> ImagePreloader', () => {
  it('should render the ImagePreloader without throwing an error', () => {
    expect(renderer.create(
      <ImagePreloader images={images} completedAction={completedAction} />,
    )).toMatchSnapshot();
  });

  it('should expect the completedAction to have been called', () => {
    setTimeout(() => expect(completedAction).toBeCalled(), 200);
  });
});
