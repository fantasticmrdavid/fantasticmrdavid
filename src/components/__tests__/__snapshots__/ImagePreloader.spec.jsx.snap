// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component -> ImagePreloader should render the ImagePreloader without throwing an error 1`] = `
<div
  className="styles__Container-sc-1i0f3dw-0 WeStR"
>
  0
  %
</div>
`;
